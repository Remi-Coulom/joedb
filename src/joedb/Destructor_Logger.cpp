#include "joedb/Destructor_Logger.h"
#include "joedb/Stream_Logger.h"

namespace joedb
{
 static Stream_Logger default_logger(std::cerr);
 Logger *Destructor_Logger::the_logger = &default_logger;

 ////////////////////////////////////////////////////////////////////////////
 void Destructor_Logger::write(const char * message)
 ////////////////////////////////////////////////////////////////////////////
 {
  the_logger->write(message);
 }

 ////////////////////////////////////////////////////////////////////////////
 void Destructor_Logger::set_logger(Logger &new_logger)
 ////////////////////////////////////////////////////////////////////////////
 {
  the_logger = &new_logger;
 }
}
