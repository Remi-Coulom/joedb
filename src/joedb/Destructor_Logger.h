#ifndef joedb_Destructor_Logger
#define joedb_Destructor_Logger

namespace joedb
{
 class Logger;

 ////////////////////////////////////////////////////////////////////////////
 class Destructor_Logger
 ////////////////////////////////////////////////////////////////////////////
 {
  private:
   static Logger *the_logger;

  public:
   static void write(const char * message) noexcept;
   static void set_logger(Logger &new_logger);
 };
}

#endif
