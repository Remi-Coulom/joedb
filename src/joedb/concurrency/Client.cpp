#include "joedb/concurrency/Client.h"

namespace joedb
{
 Client::~Client() = default;
}
