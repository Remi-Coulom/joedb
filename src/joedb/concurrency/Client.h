#ifndef joedb_Client_declared
#define joedb_Client_declared

#include "joedb/concurrency/Connection.h"
#include "joedb/concurrency/Client_Data.h"

namespace joedb
{
 ////////////////////////////////////////////////////////////////////////////
 class Client
 ////////////////////////////////////////////////////////////////////////////
 {
  private:
   //////////////////////////////////////////////////////////////////////////
   void push(bool unlock_after)
   //////////////////////////////////////////////////////////////////////////
   {
    const int64_t difference = get_checkpoint_difference();

    if (difference < 0)
     throw Exception("can't push: server is ahead of client");

    if (difference > 0)
    {
     connection.push
     (
      data.get_readonly_journal(),
      server_checkpoint,
      unlock_after
     );

     server_checkpoint =
      data.get_readonly_journal().get_checkpoint_position();
    }
    else if (unlock_after)
     connection.unlock(data.get_readonly_journal());
   }

   //////////////////////////////////////////////////////////////////////////
   void throw_if_pull_when_ahead()
   //////////////////////////////////////////////////////////////////////////
   {
    if (data.get_readonly_journal().get_position() > server_checkpoint)
     throw Exception("can't pull: client is ahead of server");
   }

  protected:
   Client_Data &data;
   Connection &connection;
   int64_t server_checkpoint;

  public:
   //////////////////////////////////////////////////////////////////////////
   Client
   //////////////////////////////////////////////////////////////////////////
   (
    Client_Data &data,
    Connection &connection
   ):
    data(data),
    connection(connection),
    server_checkpoint(connection.handshake(data.get_readonly_journal()))
   {
    data.update();
   }

   const Client_Data &get_data() const {return data;}
   Readonly_Journal &get_journal() {return data.get_readonly_journal();}
   bool is_readonly() const {return connection.is_readonly();}

   //////////////////////////////////////////////////////////////////////////
   int64_t get_checkpoint() const
   //////////////////////////////////////////////////////////////////////////
   {
    return data.get_readonly_journal().get_checkpoint_position();
   }

   //////////////////////////////////////////////////////////////////////////
   int64_t get_checkpoint_difference() const
   //////////////////////////////////////////////////////////////////////////
   {
    return
     data.get_readonly_journal().get_checkpoint_position() -
     server_checkpoint;
   }

   //////////////////////////////////////////////////////////////////////////
   void refresh_data()
   //////////////////////////////////////////////////////////////////////////
   {
    data.refresh();
   }

   //////////////////////////////////////////////////////////////////////////
   void locked_push()
   //////////////////////////////////////////////////////////////////////////
   {
    push(false);
   }

   //////////////////////////////////////////////////////////////////////////
   void push_unlock()
   //////////////////////////////////////////////////////////////////////////
   {
    push(true);
   }

   //////////////////////////////////////////////////////////////////////////
   int64_t pull()
   //////////////////////////////////////////////////////////////////////////
   {
    throw_if_pull_when_ahead();
    server_checkpoint = connection.pull(data.get_writable_journal());
    data.update();
    return server_checkpoint;
   }

   //////////////////////////////////////////////////////////////////////////
   template<typename F> void transaction(F transaction)
   //////////////////////////////////////////////////////////////////////////
   {
    throw_if_pull_when_ahead();
    server_checkpoint = connection.lock_pull(data.get_writable_journal());

    try
    {
     data.update();
     transaction(data);
     data.get_writable_journal().default_checkpoint();
    }
    catch (...)
    {
     connection.unlock(data.get_writable_journal());
     data.get_writable_journal().flush();
     throw;
    }

    push_unlock();
   }

   virtual ~Client();
 };
}

#endif
