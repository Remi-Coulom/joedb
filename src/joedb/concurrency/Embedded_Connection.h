#ifndef joedb_Embedded_Connection_declared
#define joedb_Embedded_Connection_declared

#include "joedb/concurrency/Connection.h"

namespace joedb
{
 ////////////////////////////////////////////////////////////////////////////
 class Embedded_Connection: public Connection
 ////////////////////////////////////////////////////////////////////////////
 {
  private:
   Writable_Journal &server_journal;
   bool locked;

   //////////////////////////////////////////////////////////////////////////
   int64_t handshake() final
   //////////////////////////////////////////////////////////////////////////
   {
    return server_journal.get_checkpoint_position();
   }

   //////////////////////////////////////////////////////////////////////////
   bool check_matching_content(int64_t server_checkpoint) final
   //////////////////////////////////////////////////////////////////////////
   {
    const int checkpoint = std::min
    (
     server_checkpoint,
     client_journal.get_checkpoint_position()
    );

    return
     client_journal.get_hash(checkpoint) ==
     server_journal.get_hash(checkpoint);
   }

   //////////////////////////////////////////////////////////////////////////
   void lock() final
   //////////////////////////////////////////////////////////////////////////
   {
    if (locked)
     throw Exception("Deadlock detected");
    locked = true;
   }

   //////////////////////////////////////////////////////////////////////////
   void unlock() final
   //////////////////////////////////////////////////////////////////////////
   {
    locked = false;
   }

   //////////////////////////////////////////////////////////////////////////
   int64_t pull() final
   //////////////////////////////////////////////////////////////////////////
   {
    const int64_t client_checkpoint=client_journal.get_checkpoint_position();
    const int64_t server_checkpoint=server_journal.get_checkpoint_position();

    if (client_checkpoint < server_checkpoint)
     client_journal.append_raw_tail
     (
      server_journal.get_raw_tail(client_checkpoint)
     );

    return server_checkpoint;
   }

   //////////////////////////////////////////////////////////////////////////
   void push(const int64_t server_checkpoint, bool unlock_after) final
   //////////////////////////////////////////////////////////////////////////
   {
    if (server_checkpoint != server_journal.get_checkpoint_position())
     throw Exception("pushing from bad checkpoint");

    const int64_t client_checkpoint=client_journal.get_checkpoint_position();

    if (server_checkpoint < client_checkpoint)
     server_journal.append_raw_tail
     (
      client_journal.get_raw_tail(server_checkpoint)
     );

    if (unlock_after)
     unlock();
   }

  public:
   //////////////////////////////////////////////////////////////////////////
   Embedded_Connection
   //////////////////////////////////////////////////////////////////////////
   (
    Writable_Journal &client_journal,
    Writable_Journal &server_journal
   ):
    Connection(client_journal),
    server_journal(server_journal),
    locked(false)
   {
   }
 };
}

#endif
