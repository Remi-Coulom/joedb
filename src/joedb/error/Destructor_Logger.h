#ifndef joedb_error_Destructor_Logger
#define joedb_error_Destructor_Logger

namespace joedb::error
{
 class Logger;

 ////////////////////////////////////////////////////////////////////////////
 class Destructor_Logger
 ////////////////////////////////////////////////////////////////////////////
 {
  private:
   static Logger *the_logger;

  public:
   static void write(const char * message) noexcept;
   static void set_logger(Logger *new_logger);
   static void remove_logger();
   static void set_logger();
 };
}

#endif
