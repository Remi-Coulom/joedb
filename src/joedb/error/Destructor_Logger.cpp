#include "joedb/error/Destructor_Logger.h"
#include "joedb/error/Stream_Logger.h"

#include <iostream>

namespace joedb
{
 static Stream_Logger default_logger(std::cerr);
 Logger *Destructor_Logger::the_logger = &default_logger;

 ////////////////////////////////////////////////////////////////////////////
 void Destructor_Logger::warning(std::string_view message) noexcept
 ////////////////////////////////////////////////////////////////////////////
 {
  if (the_logger)
   the_logger->write("joedb: destructor warning: " + std::string(message) + '\n');
 }

 ////////////////////////////////////////////////////////////////////////////
 void Destructor_Logger::set_logger(Logger *new_logger)
 ////////////////////////////////////////////////////////////////////////////
 {
  the_logger = new_logger;
 }

 ////////////////////////////////////////////////////////////////////////////
 void Destructor_Logger::set_logger()
 ////////////////////////////////////////////////////////////////////////////
 {
  the_logger = &default_logger;
 }

 ////////////////////////////////////////////////////////////////////////////
 void Destructor_Logger::remove_logger()
 ////////////////////////////////////////////////////////////////////////////
 {
  the_logger = nullptr;
 }
}
