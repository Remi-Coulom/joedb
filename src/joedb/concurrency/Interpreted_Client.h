#ifndef joedb_Interpreted_Client_declared
#define joedb_Interpreted_Client_declared

#include "joedb/concurrency/Client.h"
#include "joedb/interpreter/Database.h"
#include "joedb/Multiplexer.h"

namespace joedb
{
 ////////////////////////////////////////////////////////////////////////////
 class Interpreted_Client_Data: public Client_Data
 ////////////////////////////////////////////////////////////////////////////
 {
  friend class Interpreted_Client;

  private:
   Writable_Journal journal;
   Database database;
   Multiplexer multiplexer;

  public:
   Interpreted_Client_Data
   (
    joedb::Connection &connection,
    Generic_File &file
   ):
    journal(file),
    multiplexer{database, journal}
   {
   }

   Writable_Journal &get_journal() final
   {
    return journal;
   }

   const Readonly_Journal &get_journal() const final
   {
    return journal;
   }

   void update()
   {
    journal.play_until_checkpoint(database);
   }
 };

 ////////////////////////////////////////////////////////////////////////////
 class Interpreted_Client:
 ////////////////////////////////////////////////////////////////////////////
  private Interpreted_Client_Data,
  public Client
 {
  public:
   Interpreted_Client
   (
    Connection &connection,
    Generic_File &file
   ):
    Interpreted_Client_Data(connection, file),
    Client(connection, *this)
   {
   }

   const Readable &get_database() const
   {
    return database;
   }

   template<typename F> void transaction(F transaction)
   {
    Client::transaction([&]()
    {
     transaction(database, multiplexer);
    });
   }
 };
}

#endif
