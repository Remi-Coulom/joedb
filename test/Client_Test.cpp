#include "joedb/concurrency/Interpreted_Client.h"
#include "joedb/concurrency/Client.h"
#include "joedb/concurrency/Embedded_Connection.h"
#include "joedb/journal/Memory_File.h"
#include "joedb/Destructor_Logger.h"

#include "gtest/gtest.h"

using namespace joedb;

/////////////////////////////////////////////////////////////////////////////
TEST(Client, Interpreted_Client)
/////////////////////////////////////////////////////////////////////////////
{
 Memory_File server_file;
 Memory_File client1_file;
 Memory_File client2_file;

 Writable_Journal server_journal(server_file);

 using Connection = T<Embedded_Connection>;

 Interpreted_Client client1(client1_file, Connection{}, server_journal);
 Interpreted_Client client2(client2_file, Connection{}, server_journal);

 client1.transaction([](const Readable &readable, Writable &writable)
 {
  writable.create_table("person");
 });

 client1.transaction([](const Readable &readable, Writable &writable){});

 EXPECT_EQ(0, int(client2.get_database().get_tables().size()));
 client2.pull();
 EXPECT_EQ(1, int(client2.get_database().get_tables().size()));

 client2.transaction([](const Readable &readable, Writable &writable)
 {
  writable.create_table("city");
 });

 EXPECT_EQ(1, int(client1.get_database().get_tables().size()));
 client1.pull();
 EXPECT_EQ(2, int(client1.get_database().get_tables().size()));
}

/////////////////////////////////////////////////////////////////////////////
TEST(Client, Transaction_Failure)
/////////////////////////////////////////////////////////////////////////////
{
 {
  Memory_File server_file;
  Memory_File client1_file;
  Memory_File client2_file;

  Writable_Journal server_journal(server_file);

  using Connection = T<Embedded_Connection>;

  Interpreted_Client client1(client1_file, Connection{}, server_journal);
  Interpreted_Client client2(client2_file, Connection{}, server_journal);

  client1.transaction([](const Readable &readable, Writable &writable)
  {
   writable.create_table("person");
  });

  client2.pull();
  EXPECT_EQ(1, int(client2.get_database().get_tables().size()));

  try
  {
   client1.transaction([](const Readable &readable, Writable &writable)
   {
    writable.create_table("city");
    throw joedb::Exception("cancelled");
   });
   ADD_FAILURE() << "transaction should have thrown";
  }
  catch (const joedb::Exception &e)
  {
   EXPECT_STREQ(e.what(), "cancelled");
  }

  //
  // A cancelled transaction was not pushed to the server.
  //
  client2.pull();
  EXPECT_EQ(1, int(client2.get_database().get_tables().size()));

  //
  // The cancelled transaction was written locally, which prevents further
  // pulling from the server.
  //

  try
  {
   client1.transaction([](const Readable &readable, Writable &writable)
   {
    writable.create_table("country");
   });
   ADD_FAILURE() << "interrupted write should have prevented transaction";
  }
  catch (const joedb::Exception &e)
  {
   EXPECT_STREQ(e.what(), "can't pull: client is ahead of server");
  }

  client2.pull();
  EXPECT_EQ(1, int(client2.get_database().get_tables().size()));
 }
}

/////////////////////////////////////////////////////////////////////////////
TEST(Client, hash)
/////////////////////////////////////////////////////////////////////////////
{
 Memory_File server_file;
 Writable_Journal server_journal(server_file);

 {
  server_journal.create_table("person");
  server_journal.create_table("city");
  server_journal.checkpoint(Commit_Level::no_commit);
 }

 {
  Memory_File client_file;

  {
   Writable_Journal client_journal(client_file);
   client_journal.create_table("country");
   client_journal.checkpoint(Commit_Level::no_commit);
  }

  client_file.set_mode(Open_Mode::write_existing);

  try
  {
   Interpreted_Client client(client_file, T<Embedded_Connection>{}, server_journal);
   ADD_FAILURE() << "Connection with incompatible file should have failed";
  }
  catch (const joedb::Exception &e)
  {
   EXPECT_STREQ(e.what(), "Client data does not match the server");
  }
 }

 {
  Memory_File client_file;

  {
   Writable_Journal client_journal(client_file);
   client_journal.create_table("person");
   client_journal.checkpoint(Commit_Level::no_commit);
  }

  client_file.set_mode(Open_Mode::write_existing);

  try
  {
   Interpreted_Client client(client_file, T<Embedded_Connection>{}, server_journal);
   EXPECT_EQ(client.get_database().get_tables().size(), 1ULL);
   client.pull();
   EXPECT_EQ(client.get_database().get_tables().size(), 2ULL);
  }
  catch (const joedb::Exception &e)
  {
   ADD_FAILURE() << e.what();
  }
 }
}

/////////////////////////////////////////////////////////////////////////////
TEST(Client, push)
/////////////////////////////////////////////////////////////////////////////
{
 Memory_File server_file;
 Writable_Journal server_journal(server_file);

 {
  Memory_File client_file;

  //
  // Push something to the server via a connection
  //
  {
   Interpreted_Client client(client_file, T<Embedded_Connection>{}, server_journal);

   client.transaction([](const Readable &readable, Writable &writable)
   {
    writable.create_table("person");
   });
  }

  client_file.set_mode(Open_Mode::write_existing);

  //
  // Make offline modifications
  //
  {
   Writable_Journal journal(client_file);
   journal.append();
   journal.create_table("city");
   journal.checkpoint(Commit_Level::no_commit);
  }

  //
  // Connect again, and update the server
  //
  {
   Interpreted_Client client(client_file, T<Embedded_Connection>{}, server_journal);
   EXPECT_TRUE(client.get_checkpoint_difference() > 0);
   client.push_unlock();
  }
 }

 //
 // Another client connects to check the server got everything
 //
 {
  Memory_File client_file;
  Interpreted_Client client(client_file, T<Embedded_Connection>(), server_journal);
  EXPECT_TRUE(client.get_checkpoint_difference() < 0);
  client.pull();
  EXPECT_EQ(client.get_database().get_tables().size(), 2ULL);
 }
}

/////////////////////////////////////////////////////////////////////////////
TEST(Client, synchronization_error_at_handshake)
/////////////////////////////////////////////////////////////////////////////
{
 Memory_File server_file;
 Writable_Journal server_journal(server_file);

 Memory_File client_file;

 //
 // Push something to the server via a connection
 //
 {
  Interpreted_Client client(client_file, T<Embedded_Connection>{}, server_journal);
  client.transaction([](const Readable &readable, Writable &writable)
  {
   writable.create_table("person");
  });
 }

 client_file.set_mode(Open_Mode::write_existing);

 //
 // Make offline modifications
 //
 {
  Writable_Journal journal(client_file);
  journal.append();
  journal.create_table("city");
  journal.checkpoint(Commit_Level::no_commit);
 }

 //
 // Someone else makes incompatible changes
 //
 {
  Memory_File client2_file;
  Interpreted_Client client2(client2_file, T<Embedded_Connection>{}, server_journal);

  client2.transaction([](const Readable &readable, Writable &writable)
  {
   writable.create_table("company");
  });
 }

 //
 // Connect again: this should fail
 //
 try
 {
  Interpreted_Client client(client_file, T<Embedded_Connection>{}, server_journal);
  ADD_FAILURE() << "connecting should have failed\n";
 }
 catch (const Exception &e)
 {
  EXPECT_STREQ(e.what(), "Client data does not match the server");
 }
}

/////////////////////////////////////////////////////////////////////////////
TEST(Client, empty_transaction)
/////////////////////////////////////////////////////////////////////////////
{
 Memory_File server_file;
 Memory_File client_file;

 Writable_Journal server_journal(server_file);

 Interpreted_Client client(client_file, T<Embedded_Connection>{}, server_journal);

 client.transaction([](const Readable &readable, Writable &writable){});
 client.transaction([](const Readable &readable, Writable &writable){});
}
