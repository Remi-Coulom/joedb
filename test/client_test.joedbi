help
transaction
 create_table person
 add_field person name string
 insert_into person 1
 update person 1 name "Joe"
 table person
quit
push
pull
db
 table person
quit
