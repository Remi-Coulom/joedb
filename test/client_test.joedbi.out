writable_client(41)> 
General commands
~~~~~~~~~~~~~~~~
 about
 help|?
 quit
 abort
 echo on|off
 prompt on|off

Client
~~~~~~
 db
 pull [<wait_seconds>]
 pull_every [<wait_seconds>] [<sleep_seconds>]
 push
 push_every [<sleep_seconds>]

Writable Client
~~~~~~~~~~~~~~~
 transaction
 set_valid_data <true|false>
 set_timestamp <true|false>
 set_hard_checkpoint <true|false>

OK: help
writable_client(41)> writable_client(41)/joedbi> OK:  create_table person
writable_client(41)/joedbi> OK:  add_field person name string
writable_client(41)/joedbi> OK:  insert_into person 1
writable_client(41)/joedbi> OK:  update person 1 name "Joe"
writable_client(41)/joedbi> -------
  name 
-------
1 "Joe"
OK:  table person
writable_client(41)/joedbi> OK: quit
OK: transaction
writable_client(66)> OK: push
writable_client(66)> OK: pull
writable_client(66)> writable_client(66)/joedbi> -------
  name 
-------
1 "Joe"
OK:  table person
writable_client(66)/joedbi> OK: quit
OK: db
writable_client(66)> 